src/items/galleon.rs

```diff
@@ -0,0 +1,318 @@
+use macroquad::{
+    color,
+    experimental::{
+        animation::{AnimatedSprite, Animation},
+        collections::storage,
+        coroutines::{start_coroutine, Coroutine},
+        scene::{Handle, HandleUntyped, Node, RefMut},
+    },
+    prelude::*,
+    rand::gen_range,
+};
+
+use crate::{
+    capabilities,
+    components::{GunlikeAnimation, PhysicsBody, ThrowableItem},
+    nodes::Player,
+    Resources,
+};
+
+pub struct FlyingGalleon {
+    pub sprite: Texture2D,
+    pub pos: Vec2,
+    pub speed: Vec2,
+    pub facing: bool,
+    pub owner_id: u8,
+}
+
+impl FlyingGalleon {
+    pub const SPEED: f32 = 200.;
+    pub const WIDTH: f32 = 425.;
+    pub const HEIGHT: f32 = 390.;
+
+    pub fn new(owner_id: u8) -> FlyingGalleon {
+        let resources = storage::get::<Resources>();
+        let sprite = resources.items_textures["galleon/flying_galleon"];
+
+        let (pos, facing) = Self::start_position();
+        let dir = if facing {
+            Vec2::new(1., 0.)
+        } else {
+            Vec2::new(-1., 0.)
+        };
+
+        FlyingGalleon {
+            sprite,
+            pos,
+            speed: dir * Self::SPEED,
+            facing,
+            owner_id,
+        }
+    }
+
+    pub fn start_position() -> (Vec2, bool) {
+        let resources = storage::get::<Resources>();
+        let map_width =
+            resources.tiled_map.raw_tiled_map.tilewidth * resources.tiled_map.raw_tiled_map.width;
+        let map_height =
+            resources.tiled_map.raw_tiled_map.tileheight * resources.tiled_map.raw_tiled_map.height;
+
+        let facing = gen_range(0, 2) == 0;
+        let pos = Vec2::new(
+            if facing {
+                -Self::WIDTH
+            } else {
+                map_width as f32
+            },
+            gen_range(0., map_height as f32 - Self::HEIGHT),
+        );
+
+        (pos, facing)
+    }
+
+    pub fn update(&mut self) -> bool {
+        self.pos += self.speed * get_frame_time();
+
+        {
+            let resources = storage::get::<Resources>();
+            let map_width = (resources.tiled_map.raw_tiled_map.tilewidth
+                * resources.tiled_map.raw_tiled_map.width) as f32;
+            if self.pos.x + Self::WIDTH < -map_width * 0.5 || self.pos.x > map_width * 1.5 {
+                return false;
+            }
+        }
+
+        for mut player in scene::find_nodes_by_type::<crate::nodes::Player>() {
+            if player.get_hitbox().overlaps(&Rect::new(
+                self.pos.x,
+                self.pos.y,
+                Self::WIDTH,
+                Self::HEIGHT,
+            )) && player.id != self.owner_id
+                && !player.dead
+            {
+                scene::find_node_by_type::<crate::nodes::Camera>()
+                    .unwrap()
+                    .shake_noise(1.0, 10, 1.);
+                {
```

Comment:
I think braces are redundant here, unless there is some specific reason.

Reply:
Actually there is, I can't get resources without an error if I don't use them.

Reply:
This diff applies without errors:

```diff
diff --git i/src/items/galleon.rs w/src/items/galleon.rs
index dc21f74..e5ad0e3 100644
--- i/src/items/galleon.rs
+++ w/src/items/galleon.rs
@@ -94,10 +94,10 @@ pub fn update(&mut self) -> bool {
                 scene::find_node_by_type::<crate::nodes::Camera>()
                     .unwrap()
                     .shake_noise(1.0, 10, 1.);
-                {
-                    let mut resources = storage::get_mut::<Resources>();
-                    resources.hit_fxses.spawn(player.body.pos);
-                }
+
+                let mut resources = storage::get_mut::<Resources>();
+                resources.hit_fxses.spawn(player.body.pos);
+
                 let direction = self.pos.x > (player.body.pos.x + 10.);
                 player.kill(direction);
             }
```

Reply:
The error happens only when this code is run, when the player collides with the flying galleon.

Reply:
I've hit something I believe is simlar in the past. There is a MQ undocumented feature that prevents two (scene) nodes being in scope at the same time, so that the second query would return None; in that case, using braces was used to drop the first instance before the second was used. I'm checking this - there may be something similar going on.

Reply:
Very interesting - it's a borrowing problem indeed.

Specifically, `Player#kill` [queries `Resources`](https://github.com/fishfight/FishFight/blob/ff14a11ecde0a255a1bac3a1f1cdfeb0b8752640/src/nodes/player.rs#L280); if when such method is invoked, a `Resources` instance is in scope, then there will be a double borrow.

The braces cause the `Resources` instance to be dropped immediately after (fxs) usage, avoiding the problem. There is at least one `Resources` design that cleanly solves this (adding helper methods like `Resources#spawn_fx()`, which encapsulates the query), but it's outside the scope of this PR.

All good with the braces :laughing: